### PR TITLE
Support for Openshift

### PR DESCRIPTION
**Pull request by arturobrzut**
Support for Openshift
https://github.com/kubernetes-sigs/hierarchical-namespaces/issues/140

Environment Openshift 4.8
HNC v0.9.0

Hi
I install HNC on Openshift cluster 4.8
And I have such issue:
```
kubectl create namespace my-parent
namespace/my-parent created

kubectl hns create -n my-parent my-child1   
Successfully created "my-child1" subnamespace anchor in "my-parent" namespace

kubectl create namespace my-child2
namespace/my-child2 created

kubectl hns set my-child2 --parent my-parent
Setting the parent of my-child2 to my-parent

Could not update the hierarchical configuration of my-child2.
Reason: admission webhook "hierarchyconfigurations.hnc.x-k8s.io" denied the request: Cannot update hierarchy because it would overwrite the following object(s):
  * Namespace "my-child2": system:image-pullers (rbac.authorization.k8s.io/v1, Kind=RoleBinding)
  * Namespace "my-child2": system:image-builders (rbac.authorization.k8s.io/v1, Kind=RoleBinding)
  * Namespace "my-child2": system:deployers (rbac.authorization.k8s.io/v1, Kind=RoleBinding)
To fix this, please rename or remove the conflicting objects first.
```

### CODE REVIEW

File: internal/selectors/selectors.go
Comment (adrianludwin) [inline internal/selectors/selectors.go:164]: Fix comment

**Review by adrianludwin — COMMENTED** [review]
Are we sure this is the best way to do this? Does Openshift not put labels on these objects?

File: internal/selectors/selectors.go
Comment (arturobrzut) [reply internal/selectors/selectors.go:164]: no , openshift does not set labels on it
here is my output
```
oc describe rolebinding -n abc       
Name:         postgresql-operator.v0.1.1
Labels:       <none>
Annotations:  <none>
Role:
  Kind:  Role
  Name:  postgresql-operator.v0.1.1
Subjects:
  Kind            Name                 Namespace
  ----            ----                 ---------
  ServiceAccount  postgresql-operator  


Name:         system:deployers
Labels:       <none>
Annotations:  openshift.io/description:
                Allows deploymentconfigs in this namespace to rollout pods in this namespace.  It is auto-managed by a controller; remove subjects to disa...
              propagate.hnc.x-k8s.io/none: true
Role:
  Kind:  ClusterRole
  Name:  system:deployer
Subjects:
  Kind            Name      Namespace
  ----            ----      ---------
  ServiceAccount  deployer  abc


Name:         system:image-builders
Labels:       <none>
Annotations:  openshift.io/description:
                Allows builds in this namespace to push images to this namespace.  It is auto-managed by a controller; remove subjects to disable.
              propagate.hnc.x-k8s.io/none: true
Role:
  Kind:  ClusterRole
  Name:  system:image-builder
Subjects:
  Kind            Name     Namespace
  ----            ----     ---------
  ServiceAccount  builder  abc


Name:         system:image-pullers
Labels:       <none>
Annotations:  openshift.io/description:
                Allows all pods in this namespace to pull images from this namespace.  It is auto-managed by a controller; remove subjects to disable.
              propagate.hnc.x-k8s.io/none: true
Role:
  Kind:  ClusterRole
  Name:  system:image-puller
Subjects:
  Kind   Name                        Namespace
  ----   ----                        ---------
  Group  system:serviceaccounts:abc  
```

**Comment by adrianludwin** [comment]
How about checking for the presence of an annotation called `
openshift.io/description`? That seems far safer to me IMO since it's
targeted to exactly one vendor.

**Comment by arturobrzut** [comment]
It means 
not propagate all RoleBidning which contains:
  Annotations:  openshift.io/description:
?
I will recheck it

**Comment by adrianludwin** [comment]
I would go a bit further, and say: don't propagate _any_ objects with that
annotation.

**Comment by arturobrzut** [comment]
ok, it looks reasonable.
Should I update this  PR? or it will be done by you?

**Comment by adrianludwin** [comment]
Can you please update the PR? I have no ability to test on Openshift.

**Comment by arturobrzut** [comment]
ok, i will do it tomorrow

**Comment by arturobrzut** [comment]
First proposal
exclude all resources with openshift annotation, no matter what is inside this annotation

```
╭─   ~/dev/g/s/github.c/hierarchical-namespaces on   openshift-support !1 
╰─❯ oc create namespace my-parent
namespace/my-parent created
╭─   ~/dev/g/s/github.c/hierarchical-namespaces on   openshift-support !1 
╰─❯ oc create namespace my-child 
namespace/my-child created
╭─   ~/dev/g/s/github.c/hierarchical-namespaces on   openshift-support !1 
╰─❯ kubectl hns set my-child --parent my-parent
Setting the parent of my-child to my-parent
Succesfully updated 1 property of the hierarchical configuration of my-child
╭─   ~/dev/g/s/github.c/hierarchical-namespaces on   openshift-support !1 
╰─❯ oc create namespace my-child
╭─   ~/dev/g/s/github.c/hierarchical-namespaces on   openshift-support !1 
╰─❯ kubectl hns create -n my-parent second-child
Successfully created "second-child" subnamespace anchor in "my-parent" namespace
╭─   ~/dev/g/s/github.c/hierarchical-namespaces on   openshift-support !1 
╰─❯ oc hns tree  parent               
Error reading namespace parent: namespaces "parent" not found
╭─   ~/dev/g/s/github.c/hierarchical-namespaces on   openshift-support !1 
╰─❯ oc hns tree  my-parent
my-parent
├── my-child
└── [s] second-child

[s] indicates subnamespaces
╭─   ~/dev/go/src/github.com/hierarchical-namespaces on   openshift-support !1 ························································· at  10:44:53
╰─❯ ```

**Comment by adrianludwin** [comment]
/ok-to-test

Can you please squash these two commits so there's only one in the PR? Also, can you please add a note about how you tested this change, i.e. to confirm it actually works on Openshift? Finally, please add any other interesting descriptions to the CL message. A good example is here: https://github.com/kubernetes-sigs/hierarchical-namespaces/commit/485aef19c929d547503ada473c32489f13d0b5ba

@erikgb , I forgot where we left off on Friday. Are you ok with this solution for OpenShift or did you think it was critical to add the name-based resolution instead? I prefer this version for now, at least until we're able to switch a fully command-line option interface.

**Comment by adrianludwin** [comment]
/easycla

Trying to figure out how to change from the old to new CLA format

**Comment by adrianludwin** [comment]
... the presubmits failed, looks like some files are formatted incorrectly. You can use `make fmt` to fix this (do _not_ use `gofmt` as it will modify everything in `/vendor`).

**Comment by arturobrzut** [comment]
I will change it and I also work to add me to EasyCLA as corporate user

**Comment by adrianludwin** [comment]
Out of interest @arturobrzut - we're considering making an (unrelated) change to HNC that would require K8s 1.20 or later. Is that ok for you, or are you on an older version?

**Comment by arturobrzut** [comment]
it is ok

**Comment by adrianludwin** [comment]
I'm ready to merge this; I discussed it with @erikgb and while we agree that we need to stop hardcoding support for K8s distros like OpenShift, we don't want to hold up this change either.

There are only two things left:
* Can you please get the CLA authorized? (I'm guessing you're working on this? There are a bunch of bot comments at the top of this PR that might help)
* Can you please add information on how you tested this change to the commit message (not the PR message)? See https://github.com/kubernetes-sigs/hierarchical-namespaces/commit/485aef19c929d547503ada473c32489f13d0b5ba as an example.

Thanks for this change, I'm excited to see it go in!

**Comment by arturobrzut** [comment]
ok
yes i work on Easy CLA
and i will add comment for commit

**Comment by adrianludwin** [comment]
Thanks!

File: internal/selectors/selectors.go
Comment (erikgb) [inline internal/selectors/selectors.go:186]:
```suggestion
	{Key: "openshift.io/description"},
```

File: internal/selectors/selectors.go
Comment (erikgb) [inline internal/selectors/selectors.go:217]:
```suggestion
		if ok && (gotAnnotationValue == res.Value || res.Value == "") {
```

**Review by erikgb — CHANGES_REQUESTED** [review]
Some minor suggestions, but we need some tests for this new feature and some docs. Maybe use the related PR as inspiration? https://github.com/kubernetes-sigs/hierarchical-namespaces/pull/125/files

File: internal/selectors/selectors.go
Comment (adrianludwin) [reply internal/selectors/selectors.go:186]: +1

File: internal/selectors/selectors.go
Comment (adrianludwin) [reply internal/selectors/selectors.go:217]: +1

**Comment by adrianludwin** [comment]
> Some minor suggestions, but we need some tests for this new feature and some docs. Maybe use the related PR as inspiration? https://github.com/kubernetes-sigs/hierarchical-namespaces/pull/125/files

Both good points. I guess I was getting a bit lax, thanks Erik!

File: internal/selectors/selectors.go
Comment (arturobrzut) [reply internal/selectors/selectors.go:186]: Ok ,
I thought about it and you are probably right. Better to give only the name, than to have ambiguity when giving an empty annotation verse of any annotation
 I will make changes

**Comment by erikgb** [comment]
@arturobrzut You resolved both my code suggestions, but I cannot see a change in the code. Can you fix this, so we can get this merged? LGTM after the small adjustments.

**Comment by arturobrzut** [comment]
@erikgb done

**Comment by erikgb** [comment]
Code LGMT, but I think we should have some tests and docs for this.

**Comment by adrianludwin** [comment]
Actually, I can see from [this comment](https://github.com/kubernetes-sigs/hierarchical-namespaces/pull/141#issuecomment-1027761051) that @arturobrzut did test this and that it's working, so I'm happy to put this in as soon as #160 goes in first.

**Pull request merged**
